WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs

```diff
@@ -27,485 +26,299 @@ public Arena(TimeSpan period, Network network, WabiSabiConfig config, IRPCClient
-						round.LogInfo($"Number of outputs: {coinjoin.Outputs.Count}.");
-						round.LogInfo($"Serialized Size: {coinjoin.GetSerializedSize() / 1024} KB.");
-						round.LogInfo($"VSize: {coinjoin.GetVirtualSize() / 1024} KB.");
-						foreach (var (value, count) in coinjoin.GetIndistinguishableOutputs(includeSingle: false))
-						{
-							round.LogInfo($"There are {count} occurrences of {value.ToString(true, false)} BTC output.");
-						}
+					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.AliceAlreadyRegistered, $"The coin {coin.Outpoint} is already registered into round {otherAlice.Round.Id}.");
+				}
 
-						// Broadcasting.
-						await Rpc.SendRawTransactionAsync(coinjoin).ConfigureAwait(false);
-						round.WasTransactionBroadcast = true;
-						round.SetPhase(Phase.Ended);
+				try
+				{
+					var response = await round.RegisterInputAsync(alice, request, Config);
 
-						round.LogInfo($"Successfully broadcast the CoinJoin: {coinjoin.GetHash()}.");
-					}
-					else if (round.TransactionSigningStart + round.TransactionSigningTimeout < DateTimeOffset.UtcNow)
+					// Now that alice is in the round, make it available by id.
+					if (!AlicesById.TryAdd(alice.Id, alice))
 					{
-						throw new TimeoutException($"Round {round.Id}: Signing phase timed out after {round.TransactionSigningTimeout.TotalSeconds} seconds.");
+						throw new InvalidOperationException();
 					}
+
+					return response;
 				}
 				catch (Exception ex)
 				{
-					round.LogWarning($"Signing phase failed, reason: '{ex}'.");
-					await FailTransactionSigningPhaseAsync(round).ConfigureAwait(false);
+					AlicesByOutpoint.Remove(coin.Outpoint, out _);
+					throw ex;
 				}
 			}
 		}
 
-		private async Task FailTransactionSigningPhaseAsync(Round round)
+		private async Task<Coin> OutpointToCoinAsync(InputRegistrationRequest request)
 		{
-			var state = round.Assert<SigningState>();
-
-			var unsignedPrevouts = state.UnsignedInputs.ToHashSet();
+			OutPoint input = request.Input;
 
-			var alicesWhoDidntSign = round.Alices
-				.Select(alice => (Alice: alice, alice.Coin))
-				.Where(x => unsignedPrevouts.Contains(x.Coin))
-				.Select(x => x.Alice)
-				.ToHashSet();
-
-			foreach (var alice in alicesWhoDidntSign)
+			if (Prison.TryGet(input, out var inmate) && (!Config.AllowNotedInputRegistration || inmate.Punishment != Punishment.Noted))
 			{
-				Prison.Note(alice, round.Id);
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.InputBanned);
 			}
 
-			round.Alices.RemoveAll(x => alicesWhoDidntSign.Contains(x));
-			round.SetPhase(Phase.Ended);
-
-			if (round.InputCount >= Config.MinInputCountByRound)
+			var txOutResponse = await Rpc.GetTxOutAsync(input.Hash, (int)input.N, includeMempool: true).ConfigureAwait(false);
+			if (txOutResponse is null)
 			{
-				await CreateBlameRoundAsync(round).ConfigureAwait(false);
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.InputSpent);
```

Comment:
you mean if it never existed? yeah we need to bike shed some stuff.

i think a good way to do it:

1. go over all of the files & class names
2. go over error codes, shared objects etc
3. go over public api method names

there's a bunch of inconsistencies etc we should resolve. anyway, this is just moving the existing behavior around so i think it's out of scope. perhaps we should just schedule a date to do this cleanup in advance since we keep postponing it until after more important changes